ingest/expression_files/expression_files.py

```diff
@@ -93,17 +93,20 @@ def create_gene_model(
         )
 
     @staticmethod
-    def check_unique_cells(cell_names: List, study_id, client):
+    def check_unique_cells(cell_names: List, study_id, study_file_id, client):
         """Checks cell names against database to confirm matrix contains unique
             cell names.
 
+        Method acts as 'command center' for deciding what methods need to be called to check for cell uniquenes under
+            given scenarios.
+
          Parameters:
             cell_names (List[str]): List of cell names in matrix
             study_id (ObjectId): The study id the cell names belong to
+            study_file_id (ObjectId): The file id the cell names belong to
             client: MongoDB client
         """
-        COLLECTION_NAME = "data_arrays"
-        query = {
+        QUERY = {
```

Comment:
I think we have a disconnect in the logic flow here.  Previously, we only had one "bucket" of cells to check for repeats when validating that cells were unique (all expression matrices).  Now, we have two separate buckets: normalized "visualization" matrices, and raw counts matrices.  The logic for validation is the same - the only thing that we're changing is which "bucket" we use to validate against.

To achieve this, I think all that needs to happen is to append an `$in` operator for `study_file_id` to the `QUERY` object.  The contents of that array will be all expression matrices of the same "type" as the one being ingested - either raw counts, or not.  We can easily do this by getting _all_ expression matrices for the study (minus the one being ingested), and filtering out the ones that aren't of the same "type" as the matrix being ingested.

Example (this can probably be refined, but will show the logic):
```
ingest_file =  client['study_files'].find({"id": study_file_id})
# we need to check for the 'expression_file_info' object first
is_raw_counts = ingest_file['expression_file_info'] is not None && 
    ingest_file['expression_file_info']['is_raw_counts']

study_file_ids = []
study_files = client['study_files'].find({
    "study_id": study_id, 
     "$in": [
        "file_type": ['Expression Matrix', 'MM Coordinate Matrix']
     ], 
    "$not": {"id": study_file_id}
)

# grab the file id of all matrix files that are of the same type
for file in study_files:
    file_is_raw_counts = file['expression_file_info'] is not None && 
        file['expression_file_info']['is_raw_counts']
    if is_raw_count == file_is_raw_counts
        study_file_ids.append(file['id'])
        
QUERY = {
    "$and": [
        {"linear_data_type": "Study"},
        {"array_type": "cells"},
        {"study_id": study_id},
    ],
    "$nor": [{"name": "All Cells"}],
    "$in": [{"study_file_id": study_file_ids}]
}
...
```

After this, we can look for intersection as normal.

Reply:
8a03932